pep-9999.rst

```diff
@@ -0,0 +1,450 @@
+  freeze critical parts of the interpreter.
+
+As the above list shows, a side-effect of freezing an object is
+that its type becomes frozen too. Consider the following program,
+which is not legal in this PEP because it modifies the type of an
+immutable object:
+
+.. code-block:: c
+
+	class Counter:
+	  def __init__(self, initial_value):
+	    self.value = initial_value
+	  def inc(self):
+	    self.value += 1
+	  def dec(self):
+	    self.value -= 1
+	  def get(self):
+	    return self.value
+
+	c = Counter(0)
+	c.get() # returns 0 
+	freeze(c)
+	... 
+	Counter.get = lambda self: 42
+	c.get() # returns 42 
+
+Even though we froze the counter object on Line 10,
+because its class is still mutable, we are able to
+create the appearance of a change to the underlying
+object by replacing the ``get()`` method.
+
+The dangers of not freezing the type is apparent when considering
+avoiding data races in a concurrent program. If a frozen counter
+is shared between two threads, the threads are still able to
+race on the ``Counter`` class type object.
+
+When types are frozen, this problem is avoided. Note that
+freezing a class needs to freeze its superclasses as well.
+
+
+Subclassing Immutable Classes
+-----------------------------
+
+CPython classes hold references to their subclasses.
+If immutability it taken literally, it would not be
+permitted to create a subclass of an immutable type.
+Because this reference is "accidental" and does not
+get exposed to the programmer in any dangerous way,
+we permit frozen classes to be subclassed (by mutable
+classes).
+
+* Explain why and how it works
+
+
+Implementation Details
+======================
+
+1. Introduce an ``is_immutable`` flag in PyObject to indicate whether an
+   object is frozen.
+2. Modify object mutation operations (``PyObject_SetAttr``,
+   ``PyDict_SetItem``, ``PyList_SetItem``, etc.) to check the 
+   flag and raise an error when appropriate.
+3. Implement ``freeze(obj)``, ensuring it traverses object references
+   safely, including cycle detection.
+
+
+Backward Compatibility
+======================
+
+This proposal is fully backward-compatible, as no existing Python
+code will be affected unless it explicitly calls ``freeze(obj)``.
+Frozen objects will raise errors only when mutation is attempted.
+
+
+Performance Implications
+========================
+
+**TODO** The cost of checking for immutability violations is
+an extra dereference of ...
+
+
+Alternatives Considered
+=======================
+
+1. Shallow Freezing: Only mark the top-level object as immutable.
+   This would be less effective for ensuring true immutability
+   across references. In particular, this would not make it safe
+   to share the results of ``freeze()`` across threads without risking
+   data-race errors.
+2. Copy-on-Write Immutability: Instead of raising errors on
+   mutation, create a modified copy. However, this changes object
+   identity semantics and is less predictable.
```

Comment:
I think it could be worth to adding a note, that some kind of Copy-On-Write could later be implemented with a wrapper type. Similar to Rust's COW type. Personally, I would like to see something like that, also for Pyrona itself.